Simplenote/Classes/NoteOptionsViewController.swift

```diff
@@ -241,42 +302,160 @@ class NoteOptionsViewController: UITableViewController {
+            return
+        }
+
+        SPTracker.trackEditorPublishedUrlPressed()
+
+        let activityViewController = UIActivityViewController(activityItems: [publishURL],
+                                                              applicationActivities: [SPActivitySafari()])
+        presentPopover(viewController: activityViewController, from: indexPath)
     }
 
-    func handleCollaborate() {
-        ///Handle collaboration logic here
+    func handleCollaborate(from indexPath: IndexPath) {
+        SPTracker.trackEditorCollaboratorsAccessed()
+
+        let collaboratorView = SPAddCollaboratorsViewController()
+        collaboratorView.collaboratorDelegate = self
+        collaboratorView.setup(withCollaborators: note.emailTagsArray as? [String])
+
+        let navController = SPNavigationController(rootViewController: collaboratorView)
+        navController.displaysBlurEffect = true
+
+        presentPopover(viewController: navController, from: indexPath)
     }
 
     func handleMoveToTrash() {
-        ///Handle move to tash logic here
+        SPTracker.trackEditorNoteDeleted()
+        delegate?.didTapMoveToTrash(sender: self)
     }
 
     // MARK: - Navigation button handling
     @objc
     func handleDone(button: UIBarButtonItem) {
-        dismiss(animated: true)
+        dismiss(animated: true) { [popoverPresentationController] in
+            popoverPresentationController?.delegate?.popoverPresentationControllerDidDismissPopover?(popoverPresentationController!)
+        }
+    }
+
+    // MARK: - Note saving
+    func save() {
+        note.modificationDate = Date()
+        note.createPreview()
+
+        SPAppDelegate.shared().save()
+        SPTracker.trackEditorNoteEdited()
+        CSSearchableIndex.default().indexSearchableNote(note)
+    }
+
+    // MARK: - Navigation helpers
+    /// Presents a view controller as a popover from the provided index path on iPad
+    /// on iOS this will be presented modally
+    /// - Parameters:
+    ///   - viewController: The view controller to present
+    ///   - indexPath: The index path to present from
+    func presentPopover(viewController: UIViewController, from indexPath: IndexPath) {
+        if UIDevice.sp_isPad() {
+            viewController.modalPresentationStyle = .popover
+
+            let presentationController = viewController.popoverPresentationController
+            presentationController?.permittedArrowDirections = .any
+            presentationController?.sourceRect = tableView.rectForRow(at: indexPath)
+            presentationController?.sourceView = tableView
+        }
+        present(viewController, animated: true, completion: nil)
+    }
+}
+
+// MARK: - Action protocol
+//
+/// A protocol to pass row actions to the parent view controller for handling
+protocol NoteOptionsViewControllerDelegate: class {
+    func didToggleMarkdown()
+    func didTapHistory(sender: NoteOptionsViewController)
```

Comment:
@jleandroperez Do you have an opinion on delegates? I've always passed a sender of the view, just in case, but it's not actually used in the current delegate methods. My feeling here is that I can probably just remove the sender parameters

Reply:
Honestly, I try not to be biased by personal opinions!! (note the **try** word!!). But the detail is:

By exposing the internal UISwitch reference, the API would imply that it's okay for a delegate to access / tamper with the UISwitch right? (Since it's not hidden / protected anyhow).

For that reason, sending over just the property that is needed is a slightly better design, since it encapsulates the internal component, and only exposes what's 100% needed.

Reply:
Last nipticky note of the day!!!

Since the delegate needs, pretty much, a loot of visibility, would it be possible to move the protocol to the very top of the file?